### PR TITLE
WIP fix: standardize scaffold suite_test files

### PR DESCRIPTION
**Pull request by mateusoliveira43**
Scaffold `internal/controller/suite_test.go` and `internal/webhook/version/suite_test.go` have very similar code. 

This PR makes them more similar

### CODE REVIEW

File: pkg/plugins/golang/v4/scaffolds/internal/templates/webhooks/webhook_suitetest.go [hunk 1]
```diff
@@ -229,8 +229,18 @@ var _ = BeforeSuite(func() {
...
+	Expect(err).NotTo(HaveOccurred())
+
+	err = %s.AddToScheme(scheme.Scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	%s
+
 	By("bootstrapping test environment")
 	testEnv = &envtest.Environment{
+		CRDInstallOptions:     envtest.CRDInstallOptions{Scheme: scheme.Scheme},
```
Comment (mateusoliveira43) [inline pkg/plugins/golang/v4/scaffolds/internal/templates/webhooks/webhook_suitetest.go:243]: made this change because of this https://github.com/kubernetes-sigs/controller-runtime/issues/1882#issuecomment-1138880376

File: docs/book/src/cronjob-tutorial/testdata/project/internal/controller/suite_test.go [hunk 2]
```diff
@@ -56,12 +56,12 @@ Now, let's go through the code generated.
 */
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/internal/controller/suite_test.go:57]: Just a nit:

Any change that impacts the scaffolds will directly affect end users and, therefore, cannot be marked with 🌱. Please note that these emojis are used to help us generate the release notes. Both 🌱 and 📖 lead to items being listed only under the "All changes" section and not highlighted in the main release notes. While 🌱 is appropriate for test-related changes, in this case, since we're modifying the default scaffolds, it should not be used.

File: docs/book/src/cronjob-tutorial/testdata/project/internal/controller/suite_test.go [hunk 3]
```diff
@@ -74,11 +74,26 @@ var _ = BeforeSuite(func() {
 
 	ctx, cancel = context.WithCancel(context.TODO())
 
+	var err error
+	/*
+		First, the autogenerated test code will add the CronJob Kind schema to the default client-go k8s scheme.
+		This ensures that the CronJob API/Kind will be used in our test controller.
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/internal/controller/suite_test.go:80]:
```suggestion
    The CronJob Kind is added to the client-go Kubernetes scheme.
    This ensures that the CronJob API is registered with the scheme, allowing the test controller to recognize and interact with CronJob resources.
```

Add the schema does not ensure that the Kind will be used
That is required in order to allow we use it in the test.
Only the Core Types are added to the scheme by default.

File: docs/book/src/cronjob-tutorial/testdata/project/internal/controller/suite_test.go [hunk 4]
```diff
@@ -74,11 +74,26 @@ var _ = BeforeSuite(func() {
...
+	/*
+		After the schemas, you will see the following marker.
+		This marker is what allows new schemas to be added here automatically when a new API is added to the project.
+	*/
+
+	// +kubebuilder:scaffold:scheme
+
 	/*
-		First, the envtest cluster is configured to read CRDs from the CRD directory Kubebuilder scaffolds for you.
+		The envtest cluster is configured to read CRDs from the CRD directory Kubebuilder scaffolds for you.
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/internal/controller/suite_test.go:92]:
```suggestion
		/*
    The envtest environment is configured to load Custom Resource Definitions (CRDs) from the specified directory.
    This directory is scaffolded by Kubebuilder and contains the CRD manifests for the APIs defined in the project.
    The `CRDInstallOptions` ensures the specified scheme is used when installing these CRDs into the test environment.
    This setup enables the test environment to recognize and interact with the custom resources defined by these CRDs.
*/
```

The original text does not seems very accurate , I think we would need to improve it as above

File: docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go [hunk 5]
```diff
@@ -30,11 +30,7 @@ import (
 	. "github.com/onsi/gomega"
 
 	admissionv1 "k8s.io/api/admission/v1"
-
-	batchv1 "tutorial.kubebuilder.io/project/api/v1"
-
-	// +kubebuilder:scaffold:imports
-	apimachineryruntime "k8s.io/apimachinery/pkg/runtime"
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go:37]: was this one not used?
If so good catcher.

File: docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go [hunk 6]
```diff
@@ -82,29 +91,19 @@ var _ = BeforeSuite(func() {
...
 	Expect(cfg).NotTo(BeNil())
 
-	scheme := apimachineryruntime.NewScheme()
-	err = batchv1.AddToScheme(scheme)
-	Expect(err).NotTo(HaveOccurred())
-
-	err = admissionv1.AddToScheme(scheme)
-	Expect(err).NotTo(HaveOccurred())
-
-	// +kubebuilder:scaffold:scheme
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go:98]: Good catcher

File: docs/book/src/multiversion-tutorial/testdata/project/internal/webhook/v2/webhook_suite_test.go [hunk 7]
```diff
@@ -82,29 +91,19 @@ var _ = BeforeSuite(func() {
...
-	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme})
+	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme.Scheme})
 	Expect(err).NotTo(HaveOccurred())
 	Expect(k8sClient).NotTo(BeNil())
 
 	// start webhook server using Manager.
 	webhookInstallOptions := &testEnv.WebhookInstallOptions
 	mgr, err := ctrl.NewManager(cfg, ctrl.Options{
-		Scheme: scheme,
+		Scheme: scheme.Scheme,
```
Comment (camilamacedo86) [inline docs/book/src/multiversion-tutorial/testdata/project/internal/webhook/v2/webhook_suite_test.go:106]: Why this change is required?

File: docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go (hunk 5 above)
Comment (mateusoliveira43) [reply docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go:37]: it was

I removed because webhooks suite test used an "empty" scheme (with only user CRDs) and controllers suite test used a "fully loaded" scheme (with user CRDs and default objects loaded, like namespaces, pods, etc)

Without this change, users can not create a namespace in webhook envtest tests, for example

File: docs/book/src/multiversion-tutorial/testdata/project/internal/webhook/v2/webhook_suite_test.go (hunk 7 above)
Comment (mateusoliveira43) [reply docs/book/src/multiversion-tutorial/testdata/project/internal/webhook/v2/webhook_suite_test.go:106]: same motivation as this one  https://github.com/kubernetes-sigs/kubebuilder/pull/4441#discussion_r1898629715

File: docs/book/src/cronjob-tutorial/testdata/project/internal/controller/suite_test.go (hunk 2 above)
Comment (mateusoliveira43) [reply docs/book/src/cronjob-tutorial/testdata/project/internal/controller/suite_test.go:57]: updated PR title

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
HI @mateusoliveira43 

This is great work 🥇 , but I think it would be beneficial to split these changes into separate PRs, each addressing a specific context. While the changes are relatively small, having focused PRs improves clarity and makes it easier to review and track changes.

I suggest organizing the changes as follows:

- :book: Improve and fix the test information for tutorials
Focus on enhancing and correcting the tutorial content to provide clearer guidance for users.

- (go/v4): :bug: Ensure that schemes are scaffolded before the envtest starts
Address the issue where schemes need to be properly registered before the test environment is initialized to avoid potential runtime errors.

- (go/v4): Standardize the webhook and controller test layout
Introduce a consistent layout for webhook and controller tests, aligning them with best practices for readability and maintainability.

Splitting these into distinct PRs will make each change easier to review and maintain. Let me know what you think!

File: docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go [hunk 8]
```diff
@@ -67,8 +66,18 @@ var _ = BeforeSuite(func() {
...
+	Expect(err).NotTo(HaveOccurred())
+
+	err = admissionv1.AddToScheme(scheme.Scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	// +kubebuilder:scaffold:scheme
+
 	By("bootstrapping test environment")
 	testEnv = &envtest.Environment{
+		CRDInstallOptions:     envtest.CRDInstallOptions{Scheme: scheme.Scheme},
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go:80]: Why this option was added?
I do not think that we need it by default?

Comment (mateusoliveira43) [reply docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go:80]: it is needed for webhook tests https://github.com/kubernetes-sigs/kubebuilder/pull/4441#discussion_r1898614156

I think only needed for webhook, but would add to both, for consistency

File: docs/book/src/cronjob-tutorial/testdata/project/internal/controller/suite_test.go (hunk 3 above)
Comment (mateusoliveira43) [reply docs/book/src/cronjob-tutorial/testdata/project/internal/controller/suite_test.go:80]: will add to docs PR

File: docs/book/src/cronjob-tutorial/testdata/project/internal/controller/suite_test.go (hunk 4 above)
Comment (mateusoliveira43) [reply docs/book/src/cronjob-tutorial/testdata/project/internal/controller/suite_test.go:92]: will add to docs PR

File: docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go (hunk 5 above)
Comment (camilamacedo86) [reply docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go:37]: To ensure clarity and make it easier to follow the changes, I think we should split this into separate PRs for each goal. This approach will help in maintaining a clear discussion for each change, simplify the release notes, and ensure each fix or change gets the attention it deserves.

Could you please open a PR for each fix/change proposed?

Comment (mateusoliveira43) [reply docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go:37]: will do

Comment (camilamacedo86) [reply docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/webhook_suite_test.go:37]: Just to clarify: For example, in this case, we need to explain what has changed and why in the PR, and release notes so due that we cannot have many changes with diff purposes at the same PR

**Comment by mateusoliveira43** [comment]
PR was split in 3:
- https://github.com/kubernetes-sigs/kubebuilder/pull/4449
- https://github.com/kubernetes-sigs/kubebuilder/pull/4447
- https://github.com/kubernetes-sigs/kubebuilder/pull/4443
